text/0000-floating-point-additional-precision.md

```diff
@@ -0,0 +1,141 @@
+- Feature Name: `allow_extra_fp_precision`
+- Start Date: 2019-04-08
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Update the Rust specification to allow floating-point operations to provide
+*more* precision than specified, but not less precision; this allows many safe
+optimizations.
+
+# Motivation
+[motivation]: #motivation
+
+Some platforms provide instructions to run a series of floating-point
+operations quickly, such as fused multiply-add instructions; using these
+instructions can provide performance wins up to 2x or more. These instructions
+may provide *more* precision than required by IEEE floating-point operations,
+such as by doing multiple operations before rounding or losing precision.
+Similarly, high-performance floating-point code could perform multiple
+operations with higher-precision floating-point registers before converting
+back to a lower-precision format.
+
+In general, providing more precision than required should not cause a
+mathematical algorithm to fail or to lose numeric accuracy.
```

Comment:
This is incorrect. One simple counter-example is `x * x - y * y`, which is non-negative for all x and y whose squares are finite floats, but if the expression is contracted to `x.mul_add(x, - y * y)` then it can have negative results. This can of course snowball into even worse issues downstream, e.g., if this is fed into `sqrt()` to get the 2D euclidean norm, contraction can cause you to end up with NaNs on perfectly innocuous vectors.

Reply:
Any programs that have a problem with that will need to pass non-default compiler options on many common C, C++, and Fortran compilers.

That said, I'll adjust the language.

Reply:
> Any programs that have a problem with that will need to pass non-default compiler options on many common C, C++, and Fortran compilers.

_Some_ C, C++, and Fortran compilers do this (gcc, msvc), some don't (clang). If this were an universally good idea, all of them would do this, but this is not the case. That is, those languages are at most prior art, but I'm really missing from the prior art section why this would actually be a good idea. A sign change trickling down your application depending on the optimization level (or even debug-information level) can be extremely hard to debug in practice.

Reply:
> why this would actually be a good idea
> are programmers using those languages happy with that "feature"

The beginning of the RFC already makes the rationale quite clear: this allows for optimizations on the scale of 2x performance improvements, while never reducing the accuracy of a calculation compared to the mathematically accurate result.

Reply:
@rkruppe Looking again at your example, I think there's something missing from it? You said:

> One simple counter-example is x * x - y * y, which is non-negative for all x and y whose squares are finite floats

Counter-example to that: `x = 2.0`, `y = 4.0`. Both `x` and `y` square to finite floats, and `x*x - y*y` should absolutely be negative. I don't think those properties alone are enough to reasonably expect that you can call `sqrt` on that and get a non-imaginary result.

Reply:
Ugh, sorry, you're right. That's what I get for repeating the argument from memory and filling the gaps without thinking too long. In general of course x² may be smaller than y². The problematic case is only when x = y (+ aforementioned side conditions), in that case `(x * x) - (y * y)` is zero but with FMA it can be negative.

Another example, I am told, is complex multiplication when multiplying a number by its conjugate. I will not elaborate because apparently I cannot be trusted this late in the evening to work out the details correctly.

Reply:
> This is incorrect. One simple counter-example is `x * x - y * y`, which is non-negative for all x and y whose squares are finite floats, but if the expression is contracted to `x.mul_add(x, - y * y)` then it can have negative results. This can of course snowball into even worse issues downstream, e.g., if this is fed into `sqrt()` to get the 2D euclidean norm, contraction can cause you to end up with NaNs on perfectly innocuous vectors.

I suspect this is not a valid statement.

the original is in pseudocode

round64( round64(x * x) - round64(y * y) )

the contraction you gives

round64(    x * x    -  round64(y * y) )

the case for this to go negative only in the contraction case would require the
round64(x*x) to round _up_ to >= round64(y*y) while x*x itself is < round64(y*y),
so round64(x*x) == round64(y*y) by the "nearest" element of rounding; it can't cross round64(y*y)

since we're rounding to nearest, it means that x * x is equal to less than half a unit of precision away from round64(y*y).
this in turn means that x*x - round64(y*) is, while negative in this case, less than half a unit of precision way from 0, which means the outer round64() will round up to 0.

Reply:
> > This is incorrect. One simple counter-example is `x * x - y * y`, which is non-negative for all x and y whose squares are finite floats, but if the expression is contracted to `x.mul_add(x, - y * y)` then it can have negative results. This can of course snowball into even worse issues downstream, e.g., if this is fed into `sqrt()` to get the 2D euclidean norm, contraction can cause you to end up with NaNs on perfectly innocuous vectors.
> 
> I suspect this is not a valid statement.
> 
> the original is in pseudocode
> 
> round64( round64(x * x) - round64(y * y) )
> 
> the contraction you gives
> 
> round64( x * x - round64(y * y) )

If you use `y=x`, then if `round64(x*x)` rounds up, it's easy to see that `round64(x*x - round64(x*x))` is negative. This does _not_ round to zero, because units of precision are not absolute, but relative (think significant figures in scientific notation).

For reference (and more interesting floating point information!) see the "fmadd" section on https://randomascii.wordpress.com/2013/07/16/floating-point-determinism/

Reply:
So the conclusion, if I read this correctly, is that indeed increasing precision locally in some sub-computations *can* reduce precision of the overall computation, right?   (Also see [here](https://github.com/rust-lang/rfcs/pull/2686/files#r336232553).)